components/test_raftstore/src/transport_simulate.rs

```diff
@@ -162,113 +123,111 @@ impl DelayFilter {
     }
 }
 
-impl<M> Filter<M> for DelayFilter {
-    fn before(&self, _: &mut Vec<M>) -> Result<()> {
+impl Filter for DelayFilter {
+    fn before(&self, _: &mut Vec<RaftMessage>) -> Result<()> {
         thread::sleep(self.duration);
         Ok(())
     }
 }
 
-pub struct SimulateTransport<M, C: Channel<M>> {
-    filters: Arc<RwLock<Vec<Box<Filter<M>>>>>,
-    ch: Arc<Mutex<C>>,
+#[derive(Clone)]
+pub struct SimulateTransport<C> {
+    filters: Arc<RwLock<Vec<Box<Filter>>>>,
+    ch: C,
 }
 
-impl<M, C: Channel<M>> SimulateTransport<M, C> {
-    pub fn new(ch: C) -> SimulateTransport<M, C> {
+impl<C> SimulateTransport<C> {
+    pub fn new(ch: C) -> SimulateTransport<C> {
         SimulateTransport {
             filters: Arc::new(RwLock::new(vec![])),
-            ch: Arc::new(Mutex::new(ch)),
+            ch,
         }
     }
 
     pub fn clear_filters(&mut self) {
         self.filters.wl().clear();
     }
 
-    pub fn add_filter(&mut self, filter: Box<Filter<M>>) {
+    pub fn add_filter(&mut self, filter: Box<Filter>) {
         self.filters.wl().push(filter);
     }
 }
 
-impl<M, C: Channel<M>> Channel<M> for SimulateTransport<M, C> {
-    fn send(&self, msg: M) -> Result<()> {
-        let mut taken = 0;
-        let mut msgs = vec![msg];
-        let filters = self.filters.rl();
-        let mut res = Ok(());
-        for filter in filters.iter() {
-            taken += 1;
-            res = filter.before(&mut msgs);
+fn filter_send<H>(filters: &Arc<RwLock<Vec<Box<Filter>>>>, msg: RaftMessage, mut h: H) -> Result<()>
+where
+    H: FnMut(RaftMessage) -> Result<()>,
+{
+    let mut taken = 0;
+    let mut msgs = vec![msg];
+    let filters = filters.rl();
+    let mut res = Ok(());
+    for filter in filters.iter() {
+        taken += 1;
+        res = filter.before(&mut msgs);
+        if res.is_err() {
+            break;
+        }
+    }
+    if res.is_ok() {
+        for msg in msgs {
+            res = h(msg);
             if res.is_err() {
                 break;
             }
         }
-        if res.is_ok() {
-            for msg in msgs {
-                res = self.ch.lock().unwrap().send(msg);
-                if res.is_err() {
-                    break;
-                }
-            }
-        }
-        for filter in filters[..taken].iter().rev() {
-            res = filter.after(res);
-        }
-        res
     }
-}
-
-impl<M, C: Channel<M>> Clone for SimulateTransport<M, C> {
-    fn clone(&self) -> SimulateTransport<M, C> {
-        SimulateTransport {
-            filters: Arc::clone(&self.filters),
-            ch: Arc::clone(&self.ch),
-        }
+    for filter in filters[..taken].iter().rev() {
+        res = filter.after(res);
```

Comment:
If B experiences an error, C should not be called. That's why there is a `..taken` slice.

Reply:
Right, I caught that. But I mean in the `after();` call which returns a `Result`. We don't check to see if the outcome is an error, we just keep looping.

Reply:
Yes. It's intended. Because all filters that have been called `before` should also be called `after`. So all of them can decide to return errors or not. So even a preceding filter returns error, the filter right behind it can still have chance to recover from it.